### ISSUE TITLE
NTI not reporting wrong arg count when namespace is not const

### ISSUE THREAD

**Issue by shicks** [google/closure-compiler#2742]
[repro]

```js
let ns = {};

/** @param {number} required */
ns.func = function(required) {}

function f() {
  ns.func();
}
```

This should generate an error.  Indeed, if `ns` is declared as `const` then it does:

```
input0:7: WARNING - Function ns.func: called with 0 argument(s). Function requires at least 1 argument(s) and at most 1.
  ns.func();
  ^^^^^^^^^
```

The fact that the function call in inside an inner function is important.  Without the indirection, NTI successfully detects the error.  I'm guessing this has something to do with the function being added differently when `ns` is just an object (as a stray prop?) vs a legit namespace, and in the former case, the function ends up as `?` when it's closed over.

I don't know what the best solution would be.  We've had lots of problems from namespaces not being declared `const`, so early detection might be good there.  Maybe recognizing "effectively-const" at an earlier stage would be sufficient?  Maybe reporting an error if a non-const object gets a method assigned to it with jsdoc (though this would probably break a lot, I could imagine it has valid use cases, and it still might miss errors)?

@dimvar might have some thoughts here.

[repro]: https://closure-compiler-debugger.appspot.com/#input0%3Dlet%2520ns%2520%253D%2520%257B%257D%253B%250A%250A%252F**%2520%2540param%2520%257Bnumber%257D%2520required%2520*%252F%250Ans.func%2520%253D%2520function(required)%2520%257B%257D%250A%250Afunction%2520f()%2520%257B%250A%2520%2520ns.func()%253B%250A%257D%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3Dtrue%26CHECK_SYMBOLS%3Dtrue%26MISSING_PROPERTIES%3Dtrue%26TRANSPILE%3Dtrue%26CHECK_TYPES%3Dtrue%26CHECK_TYPES_NEW_INFERENCE%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue
